### DISCUSSION TITLE
Execute job in multiprocessing in Python

### DISCUSSION THREAD

**Discussion by dagsir** [dagster-io/dagster#8467] [Q&A] [answered]
Is there a way to kick off a local Dagster job within a Python program and still use Python multiprocessing?

We have some integration tests that do one or more pipeline runs, as well as some CLI entry points, all of which I'd like to preserve while migrating to Dagster. I found `DagsterInstance` but it looks like it's for internal use. What I'm doing now is providing a wrapper like this:

```python
def run_pipeline(...):
    with tempfile.NamedTemporaryFile(mode="w", suffix=".yaml") as temp_config_file:
        shared_config = {
            "a_config_option": x,
            "another_config_option": y,
        }
        yaml.dump(
            {"resources": {"shared_config": {"config": shared_config}}},
            temp_config_file,
        )
        temp_config_file.flush()

        subprocess.check_call([
            "dagster",
            "job",
            "execute",
            "-m", "my.dagster_repo.module",
            "-j", "my_job_name",
            "-c", temp_config_file.name,
        ])
```
It would be more convenient to just call this directly in Python. Ideally including `dagster.yaml` settings too would be great, so the wrapper script can set some of those options.

---

*The question was originally asked in [Dagster Slack](https://dagster.slack.com/archives/C01U954MEER/p1646321401698059?thread_ts=1646321401.698059&cid=C01U954MEER).*

**Comment by yuhan** [comment, marked as answer]
To execute a job with multiprocessing in Python, you can use the `execute_job` API and wrap the job using [`reconstructable`](https://docs.dagster.io/_apidocs/jobs#reconstructable-jobs) or [`build_reconstructable_job`](https://docs.dagster.io/_apidocs/jobs#dagster.build_reconstructable_job), such as:
```python
from dagster import execute_job, reconstructable, DagsterInstance

result = execute_job(
    reconstructable(my_job_name),
    instance=DagsterInstance.get(),
)
```

Btw, if you'd like to run the job sequentially in process, you can simply call `my_job.execute_in_process()`. More details in this [doc](https://docs.dagster.io/concepts/ops-jobs-graphs/job-execution#python-apis).

**Comment by sryza** [comment]
Here's an issue for tracking an API that makes it possible to do this for assets: https://github.com/dagster-io/dagster/issues/16236
